### ISSUE TITLE
Possible regression of version 18.1.2?

### ISSUE THREAD

**Issue by hanmingliu2** [llvm/llvm-project#86009]
Congratulations on releasing version 18.1.2. Fantastic work! However, when upgrading my project to using version 18.1.2 of the LLVM toolchain, my project failed to compile. Interestingly, when version 17.0.6 was used, it compiled and ran without any issue. The full stack trace is a bit too long so I've included the portion that I think are useful. Let me know if the full stack trace is needed.

```
In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vector:308:
In file included from /usr/local/llvm/include/c++/v1/__algorithm/copy.h:12:
In file included from /usr/local/llvm/include/c++/v1/__algorithm/copy_move_common.h:18:
In file included from /usr/local/llvm/include/c++/v1/__string/constexpr_c_functions.h:14:
In file included from /usr/local/llvm/include/c++/v1/__memory/construct_at.h:23:
In file included from /usr/local/llvm/include/c++/v1/new:364:
In file included from /usr/local/llvm/include/c++/v1/cstdlib:87:
/usr/local/llvm/include/c++/v1/stdlib.h:145:30: error: unknown type name 'ldiv_t'
  145 | inline _LIBCPP_HIDE_FROM_ABI ldiv_t div(long __x, long __y) _NOEXCEPT { return ::ldiv(__x, __y); }
      |                              ^
/usr/local/llvm/include/c++/v1/stdlib.h:145:82: error: no member named 'ldiv' in the global namespace
  145 | inline _LIBCPP_HIDE_FROM_ABI ldiv_t div(long __x, long __y) _NOEXCEPT { return ::ldiv(__x, __y); }
      |                                                                                ~~^
/usr/local/llvm/include/c++/v1/stdlib.h:147:30: error: unknown type name 'lldiv_t'
  147 | inline _LIBCPP_HIDE_FROM_ABI lldiv_t div(long long __x, long long __y) _NOEXCEPT { return ::lldiv(__x, __y); }
      |                              ^
/usr/local/llvm/include/c++/v1/stdlib.h:147:93: error: no member named 'lldiv' in the global namespace
  147 | inline _LIBCPP_HIDE_FROM_ABI lldiv_t div(long long __x, long long __y) _NOEXCEPT { return ::lldiv(__x, __y); }
      |                                                                                           ~~^
In file included from godot-cpp/src/godot.cpp:33:
In file included from godot-cpp/include/godot_cpp/classes/editor_plugin_registration.hpp:34:
In file included from godot-cpp/include/godot_cpp/templates/vector.hpp:39:
In file included from godot-cpp/include/godot_cpp/core/error_macros.hpp:36:
In file included from /usr/local/llvm/include/c++/v1/atomic:591:
In file included from /usr/local/llvm/include/c++/v1/__atomic/aliases.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_base.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_sync.h:19:
In file included from /usr/local/llvm/include/c++/v1/__thread/poll_with_backoff.h:15:
In file included from /usr/local/llvm/include/c++/v1/__chrono/high_resolution_clock.h:14:
/usr/local/llvm/include/c++/v1/__chrono/system_clock.h:35:10: error: reference to unresolved using declaration
   35 |   static time_t to_time_t(const time_point& __t) _NOEXCEPT;
      |          ^
/usr/local/llvm/include/c++/v1/ctime:67:1: note: using declaration annotated with 'using_if_exists' here
   67 | using ::time_t _LIBCPP_USING_IF_EXISTS;
      | ^
In file included from godot-cpp/src/godot.cpp:33:
In file included from godot-cpp/include/godot_cpp/classes/editor_plugin_registration.hpp:34:
In file included from godot-cpp/include/godot_cpp/templates/vector.hpp:39:
In file included from godot-cpp/include/godot_cpp/core/error_macros.hpp:36:
In file included from /usr/local/llvm/include/c++/v1/atomic:591:
In file included from /usr/local/llvm/include/c++/v1/__atomic/aliases.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_base.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_sync.h:19:
In file included from /usr/local/llvm/include/c++/v1/__thread/poll_with_backoff.h:15:
In file included from /usr/local/llvm/include/c++/v1/__chrono/high_resolution_clock.h:14:
/usr/local/llvm/include/c++/v1/__chrono/system_clock.h:36:33: error: reference to unresolved using declaration
   36 |   static time_point from_time_t(time_t __t) _NOEXCEPT;
      |                                 ^
/usr/local/llvm/include/c++/v1/ctime:67:1: note: using declaration annotated with 'using_if_exists' here
   67 | using ::time_t _LIBCPP_USING_IF_EXISTS;
      | ^
In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vector:314:
In file included from /usr/local/llvm/include/c++/v1/__algorithm/remove.h:12:
In file included from /usr/local/llvm/include/c++/v1/__algorithm/find.h:31:
In file included from /usr/local/llvm/include/c++/v1/cwchar:114:
In file included from /usr/local/llvm/include/c++/v1/wchar.h:129:
/usr/local/llvm/include/c++/v1/__mbstate_t.h:51:4: error: "We don't know how to get the definition of mbstate_t without <wchar.h> on your platform."
   51 | #  error "We don't know how to get the definition of mbstate_t without <wchar.h> on your platform."
      |    ^
In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vector:314:
In file included from /usr/local/llvm/include/c++/v1/__algorithm/remove.h:12:
In file included from /usr/local/llvm/include/c++/v1/__algorithm/find.h:31:
In file included from /usr/local/llvm/include/c++/v1/cwchar:114:
/usr/local/llvm/include/c++/v1/wchar.h:148:20: error: use of undeclared identifier 'wcschr'
  148 |   return (wchar_t*)wcschr(__s, __c);
      |                    ^
/usr/local/llvm/include/c++/v1/wchar.h:158:20: error: use of undeclared identifier 'wcspbrk'
  158 |   return (wchar_t*)wcspbrk(__s1, __s2);
      |                    ^
/usr/local/llvm/include/c++/v1/wchar.h:169:20: error: use of undeclared identifier 'wcsrchr'; did you mean 'wcschr'?
  169 |   return (wchar_t*)wcsrchr(__s, __c);
      |                    ^
/usr/local/llvm/include/c++/v1/wchar.h:150:72: note: 'wcschr' declared here
  150 | inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_PREFERRED_OVERLOAD const wchar_t* wcschr(const wchar_t* __s, wchar_t __c) {
      |                                                                        ^
/usr/local/llvm/include/c++/v1/wchar.h:179:20: error: use of undeclared identifier 'wcsstr'; did you mean 'wcschr'?
  179 |   return (wchar_t*)wcsstr(__s1, __s2);
      |                    ^
/usr/local/llvm/include/c++/v1/wchar.h:153:66: note: 'wcschr' declared here
  153 | inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_PREFERRED_OVERLOAD wchar_t* wcschr(wchar_t* __s, wchar_t __c) {
      |                                                                  ^
/usr/local/llvm/include/c++/v1/wchar.h:179:20: error: no matching function for call to 'wcschr'
  153 |   return (wchar_t*)wcsstr(__s1, __s2);
      |                    ^
/usr/local/llvm/include/c++/v1/wchar.h:153:66: note: candidate disabled: <no message provided>
  153 | inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_PREFERRED_OVERLOAD wchar_t* wcschr(wchar_t* __s, wchar_t __c) {
      |                              ~~~~~~~~~~~~~~~~~~~~~~~~~~          ^
/usr/local/llvm/include/c++/v1/wchar.h:179:27: error: cannot initialize a parameter of type 'wchar_t *' with an lvalue of type 'const wchar_t *'
  179 |   return (wchar_t*)wcsstr(__s1, __s2);
      |                           ^~~~
/usr/local/llvm/include/c++/v1/wchar.h:153:82: note: passing argument to parameter '__s' here
  153 | inline _LIBCPP_HIDE_FROM_ABI _LIBCPP_PREFERRED_OVERLOAD wchar_t* wcschr(wchar_t* __s, wchar_t __c) {
      |                                                                                  ^
/usr/local/llvm/include/c++/v1/wchar.h:190:20: error: use of undeclared identifier 'wmemchr'
  190 |   return (wchar_t*)wmemchr(__s, __c, __n);
      |                    ^
In file included from godot-cpp/src/classes/editor_plugin_registration.cpp:31:
In file included from godot-cpp/include/godot_cpp/classes/editor_plugin_registration.hpp:34:
In file included from godot-cpp/include/godot_cpp/templates/vector.hpp:39:
In file included from godot-cpp/include/godot_cpp/core/error_macros.hpp:36:
In file included from /usr/local/llvm/include/c++/v1/atomic:591:
In file included from /usr/local/llvm/include/c++/v1/__atomic/aliases.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_base.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_sync.h:19:
In file included from /usr/local/llvm/include/c++/v1/__thread/poll_with_backoff.h:15:
In file included from /usr/local/llvm/include/c++/v1/__chrono/high_resolution_clock.h:14:
/usr/local/llvm/include/c++/v1/__chrono/system_clock.h:35:10: error: reference to unresolved using declaration
   35 |   static time_t toIn file included from godot-cpp/src/classes/low_level.cpp:31:
In file included from godot-cpp/gen/include/godot_cpp/classes/file_access.hpp:37:
In file included from godot-cpp/gen/include/godot_cpp/variant/packed_byte_array.hpp:38:
In file included from godot-cpp/include/godot_cpp/core/error_macros.hpp:36:
In file included from /usr/local/llvm/include/c++/v1/atomic:591:
In file included from /usr/local/llvm/include/c++/v1/__atomic/aliases.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_base.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_sync.h:19:
In file included from /usr/local/llvm/include/c++/v1/__thread/poll_with_backoff.h:15:
In file included from /usr/local/llvm/include/c++/v1/__chrono/high_resolution_clock.h:14:
/usr/local/llvm/include/c++/v1/__chrono/system_clock.h:35:10: error: reference to unresolved using declaration
   35 |   static time_t to_time_t(const t_time_t(const time_point& __t) _NOEXCEPT;
      |          ^
/usr/local/llvm/include/c++/v1/ctime:67:1: note: using declaration annotated with 'using_if_exists' here
   67 | using ::time_t _LIBCPP_USING_IF_EXISTS;
      | ^
iIn file included from me_poingodot-cpp/src/classes/editor_plugin_registration.cpp:31:
In file included from godot-cpp/include/godot_cpp/classes/editor_plugin_registration.hpp:34:
In file included from godot-cpp/include/godot_cpp/templates/vector.hpp:39:
In file included from godot-cpp/include/godot_cpp/core/error_macros.hpp:36:
In file included from /usr/local/llvm/include/c++/v1/atomic:591:
In file included from /usr/local/llvm/include/c++/v1/__atomic/aliases.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic.h:12:
t& __t) _NOEXCEPT;
      |          ^
/usr/local/llvm/include/c++/v1/ctime:67:1: note: using declaration annotated with 'using_if_exists' here
   67 | using ::time_t _LIBCPP_USING_IF_EXISTS;
      | ^
In file included from godot-cpp/src/classes/low_level.cpp:31:
In file included from godot-cpp/gen/include/godot_cpp/classes/file_access.hpp:37:
In file included from godot-cpp/gen/include/godot_cpp/variant/packed_byte_array.hpp:38:
In file included from godot-cpp/include/godot_cpp/core/error_macros.hpp:36:
In file included from /usr/local/llvm/include/c++/v1/atomic:591:
In file included from /usr/local/llvm/include/c++/v1/__atomic/aliases.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_base.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_sync.h:19:
In file included from /usr/local/llvm/include/c++/v1/__thread/poll_with_backoff.h:15:
In file included from /usr/local/llvm/include/c++/v1/__chrono/high_resolution_clock.h:14:
/usr/local/llvm/include/c++/v1/__chrono/system_clock.h:36:33: error: reference to unresolved using declaration
   36 |   static time_point from_time_t(time_t __In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_base.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_sync.h:19:
In file included from /usr/local/llvm/include/c++/v1/__thread/poll_with_backoff.h:15:
In file included from /usr/local/llvm/include/c++/v1/__chrono/high_resolution_clock.h:14:
/usr/local/llvm/include/c++/v1/__chrono/system_clock.h:36:33: In file included from error: godot-cpp/src/core/memory.cpp:31:
In file included from godot-cpp/include/godot_cpp/core/memory.hpp:38:
In file included from godot-cpp/include/godot_cpp/core/error_macros.hpp:36:
reference to unresolved using declarationt) _NIn file included from OEXCEPT/usr/local/llvm/include/c++/v1/atomic;:591:
In file included from /usr/local/llvm/include/c++/v1/__atomic/aliases.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_base.h
:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_sync.h:19:
In file included from /usr/local/llvm/include/c++/v1/__thread/poll_with_backoff.h   :3615 | :
 In file included from  /usr/local/llvm/include/c++/v1/__chrono/high_resolution_clock.hs:t14a:
t/usr/local/llvm/include/c++/v1/__chrono/system_clock.hi:c35 :t10i:m eerror: _reference to unresolved using declarationp
oint fr
om   _35t | i  static       |                                 ^
time_t /usr/local/llvm/include/c++/v1/ctimet:o67_:t1i:m enote: _using declaration annotated with 'using_if_exists' heret
(const time_point& __t) _NOEXCEPT;
      |          ^
/usr/local/llvm/include/c++/v1/ctime:67:1: note: using declaration annotated with 'using_if_exists' here
   67 | usin   67 | g ::time_t _LIBCPP_USING_IF_EusXISTS;
      | ^
ing ::time_t _LIBCPP_USING_IF_EXISIn file included from godot-cpp/src/core/memory.cppTS;:31:
In file included from godot-cpp/include/godot_cpp/core/memory.hpp:38:
In file included from godot-cpp/include/godot_cpp/core/error_macros.hpp:36:
In file included from /usr/local/llvm/include/c++/v1/atomic:591:
In file included from /usr/local/llvm/include/c++/v1/__atomic/aliases.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic.h:12
      | ^
:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_base.h:12:
In file included from /usr/local/llvm/include/c++/v1/__atomic/atomic_sync.h:19:
In file included from /usr/local/llvm/include/c++/v1/__thread/poll_with_backoff.h:15:
In file included from /usr/local/llvm/include/c++/v1/__chrono/high_resolution_clock.h:14:
/usr/local/llvm/include/c++/v1/__chrono/system_clock.h:36:33: error: reference to unresolved using declaration
   36 |   static time_point from_time_t(time_t __t) _NOEXCEPT;
      |                                 ^
/usr/local/llvm/include/c++/v1/ctime:67:1: note: using declaration annotated with 'using_if_exists' here
   67 | using ::time_t _LIBCPP_USING_In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vector:325:
In file included from /usr/local/llvm/include/c++/v1/__format/formatter_bool.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/concepts.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/format_parse_context.h:14:
In file included from /usr/local/llvm/include/c++/v1/__format/format_error.h:15:
In file included from /usr/local/llvm/include/c++/v1/stdexcept:283:
In file included from /usr/local/llvm/include/c++/v1/exception:82:
/usr/local/llvm/include/c++/v1/__exception/exception_ptr.h:111:8: error: reference to unresolved using declaration
  111 |   std::abort();
      |        ^
/usr/local/llvm/include/c++/v1/cstdlib:124:1: note: using declaration annotated with 'using_if_exists' here
  124 | using ::abort _LIBCPP_USING_IF_EXISTS;
      | ^
IF_EXISTS;
      | ^
me_t(time_t __t) _NOEXCEPT;
      |                                 ^
/usr/local/llvm/include/c++/v1/ctime:67:1: note: using declaration annotated with 'using_if_exists' here
   67 | using ::time_t _LIBCPP_USING_IF_EXISTS;
      | ^
In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vector:325:
In file included from /usr/local/llvm/include/c++/v1/__format/formatter_bool.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/concepts.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/format_parse_context.h:14:
In file included from /usr/local/llvm/include/c++/v1/__format/format_error.h:15:
In file included from /usr/local/llvm/include/c++/v1/stdexcept:284:
/usr/local/llvm/include/c++/v1/iosfwd:136:14: error: reference to unresolved using declaration
  136 | typedef fpos<mbstate_t> streampos;
      |              ^
/usr/local/llvm/include/c++/v1/__std_mbstate_t.h:25:1: note: using declaration annotated with 'using_if_exists' here
   25 | using ::mbstate_t _LIBCPP_USING_IF_EXISTS;
      | ^
In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vector:325:
In file included from /usr/local/llvm/include/c++/v1/__format/formatter_bool.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/concepts.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/format_parse_context.h:14:
In file included from /usr/local/llvm/include/c++/v1/__format/format_error.h:15:
In file included from /usr/local/llvm/include/c++/v1/stdexcept:284:
/usr/local/llvm/include/c++/v1/iosfwd:136:14: error: reference to unresolved using declaration
  136 | typedef fpos<mbstate_t> streampos;
      |              ^
/usr/local/llvm/include/c++/v1/__std_mbstate_t.h:25:1: note: using declaration annotated with 'using_if_exists' here
   25 | using ::mbstate_t _LIBCPP_USING_IF_EXISTS;
      | ^
In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vector:325:
In file included from /usr/local/llvm/include/c++/v1/__format/formatter_bool.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/concepts.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/format_parse_context.h:14:
In file included from /usr/local/llvm/include/c++/v1/__format/format_error.h:15:
In file included from /usr/local/llvm/include/c++/v1/stdexcept:284:
/usr/local/llvm/include/c++/v1/iosfwd:138:14: error: reference to unresolved using declaration
  138 | typedef fpos<mbstatIn file included from godot-cpp/src/core/method_bind.cpp:31:
In file included from godot-cpp/include/godot_cpp/core/method_bind.hpp:34:
In file included from godot-cpp/include/godot_cpp/core/binder_common.hpp:36:
In file included from godot-cpp/include/godot_cpp/core/method_ptrcall.hpp:36:
In file included from godot-cpp/include/godot_cpp/core/object.hpp:38:
In file included from godot-cpp/include/godot_cpp/core/property_info.hpp:38:
In file included from godot-cpp/include/godot_cpp/varie_t> wstreampos;
      |              ^
/usr/local/llvm/include/c++/v1/__std_mbstate_t.h:25:1: note: using declaration annotated with 'using_if_exists' here
   25 | using ::mbstate_t _LIBCPP_USING_IF_EXISTS;
      | ^
In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vector:325:
In file included from /usr/local/llvm/include/c++/v1/__format/formatter_bool.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/concepts.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/format_parse_context.h:14:
In file included from /usr/local/llvm/include/c++/v1/__format/format_error.h:15:
In file included from /usr/local/llvm/include/c++/v1/stdexcept:284:
/usr/local/llvm/include/c++/v1/iosfwd:138:14: error: reference to unresolved using declaration
  138 | typedef fpos<mbstate_t> wstreampos;
      |              ^
/usr/local/llvm/include/c++/v1/__std_mbstate_t.h:25:1: note: using declaration annotated with 'using_if_exists' here
   25 | using ::mbstate_t _LIBCPP_USING_IF_EXISTS;
      | ^
In file included from godot-cpp/src/classes/wrapped.cpp:31:
In file included from /usr/local/llvm/include/c++/v1/vectorant/variant.hpp:325:
:36In file included from /usr/local/llvm/include/c++/v1/__format/formatter_bool.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/concepts.h:17:
In file included from /usr/local/llvm/include/c++/v1/__format/format_parse_context.h:14:
In file included from /usr/local/llvm/include/c++/v1/__format/format_error.h:15:
In file included from /usr/local/llvm/include/c++/v1/stdexcept:284:
/usr/local/llvm/include/c++/v1/iosfwd:143:14: error: reference to unresolved using declaration
  143 | typedef fpos<mbstate_t> u16streampos;
      |              ^
/usr/local/llvm/include/c++/v1/__std_mbstate_t.h:25:1: note: using declaration annotated with 'using_if_exists' here
   25 | using ::mbstate_t _LIBCPP_USING_IF_EXIST:
```

As `mbstate_t` appeared in the stack trace, could this be possibly related to #84884 ? FYI - the toolchain is built from source on a M1 Mac running macOS 14.4 with the following CMake configuration:
```
cmake ../llvm -G Ninja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld;lldb" -DLLVM_ENABLE_RUNTIMES="compiler-rt;libcxx;libcxxabi;libunwind" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr/local/llvm -DLLVM_ENABLE_Z3_SOLVER=ON -DLLVM_INCLUDE_BENCHMARKS=OFF -DLLVM_INCLUDE_EXAMPLES=OFF -DLLVM_INCLUDE_TESTS=OFF -DLLVM_TARGETS_TO_BUILD="AArch64;X86" -DLLDB_USE_SYSTEM_DEBUGSERVER=ON
```

Any advice or pointer will be greatly appreciated!

**Comment by mordante** [comment]
Thanks for the report. Could you test whether reverting a595b931f1f91897317a4257df313bddfeb029a6 fixes it for you? @petrhosek can you test whether reverting solves https://github.com/llvm/llvm-project/issues/84884?

**Comment by ldionne** [comment]
Can you show us how you are compiling `wrapped.cpp` (the compiler command-line) and its contents?

**Comment by hanmingliu2** [comment]
Hi @ldionne , thank you the quick follow up! My project uses the open-source Godot Engine. Here is the content of [wrapped.cpp](https://github.com/godotengine/godot-cpp/blob/godot-4.2.1-stable/src/classes/wrapped.cpp). Based on the stack trace, the line that caused the error is a simple statement of `#include <vector>`.

**Comment by hanmingliu2** [comment]
@ldionne Sorry for forgetting about the compiler command line. Here it is: 
```
clang++ -o godot-cpp/src/classes/wrapped.macos.template_debug.universal.o -c -nostdinc++ -nostdlib++ -fdata-sections -ffunction-sections -std=c++17 -fno-exceptions -arch x86_64 -arch arm64 -fvisibility=hidden -gdwarf-4 -g2 -O2 -I/opt/homebrew/opt/llvm/include/c++/v1 godot-cpp/src/classes/wrapped.cpp
```

**Comment by mordante** [comment]
This looks like part of the real issue
```
In file included from /usr/local/llvm/include/c++/v1/wchar.h:129:
/usr/local/llvm/include/c++/v1/__mbstate_t.h:51:4: error: "We don't know how to get the definition of mbstate_t without <wchar.h> on your platform."
   51 | #  error "We don't know how to get the definition of mbstate_t without <wchar.h> on your platform."
      |    ^
```
Can you show the full compiler command being executed? This should show the Clang added include paths.

**Comment by hanmingliu2** [comment]
Hi @mordante , I used the `-v` option to show complete compiler output. Here are the Clang added include paths:
```
ignoring nonexistent directory "/usr/local/include"
ignoring nonexistent directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/llvm/include/c++/v1
 godot-cpp/gdextension
 godot-cpp/include
 godot-cpp/gen/include
 /usr/local/llvm/lib/clang/18/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
```

**Comment by mordante** [comment]
Thanks! I expect only the first path ` /usr/local/llvm/include/c++/v1` contains a `wchar.h` and none of the other paths. Is that correct?

I don't know what the path on Apple is where the C library headers are available. @ldionne is much more familiar with the Apple platform than I am.

**Comment by hanmingliu2** [comment]
@mordante Thank you for explaining! Your expectation is correct. I actually found the `wchar.h` in the `/usr/local/llvm/include/c++/v1` directory. I don't understand why Clang/Clang++ cannot find it...

**Comment by mordante** [comment]
> @mordante Thank you for explaining! Your expectation is correct. I actually found the `wchar.h` in the `/usr/local/llvm/include/c++/v1` directory. I don't understand why Clang/Clang++ cannot find it...

It can find that `wchar.h`. That header is libc++'s `wchar.h` which depends on the `wchar.h` provided by the system's libc. On my Linux system that is `/usr/include/wchar.h`, however I don't know where Apple stores its system libc headers.

**Comment by hanmingliu2** [comment]
Thank you for clarifying! To figure out why, I compared the full include paths between version 17.0.6 and version 18.1.2.

**Version 17.0.6**
```
 /opt/homebrew/opt/llvm/include/c++/v1
 godot-cpp/gdextension
 godot-cpp/include
 godot-cpp/gen/include
 /opt/homebrew/Cellar/llvm/17.0.6_1/lib/clang/17/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include
 /Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/System/Library/Frameworks (framework directory)
```

**Version 18.1.2**
```
 /usr/local/llvm/include/c++/v1
 godot-cpp/gdextension
 godot-cpp/include
 godot-cpp/gen/include
 /usr/local/llvm/lib/clang/18/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
```

Notice how version 18.1.2 included two extra paths at the end **instead of the macOS SDK path**. I believe the **macOS SDK path** is where the Apple libc headers and their shared libraries are. From my prior experiences, they are usually included as `isysroot` by compilers by default. However, that's not the case for version 18.1.2 of Clang/Clang++.

To test my theory, I actually added `-isysroot <macOS SDK path>` as both a compiler flag and a link flag. Then my project built without an issue!

Hope this helps! Lastly, I want to thank you both again for your assistance!

**Comment by ldionne** [comment]
I don't think this is a difference between `17.0.6` and `18.1.2`. I think this is a difference between how the toolchain was configured and built in both cases. Your `17.0.6` version seems to have been installed via Homebrew and configured correctly to have a default sysroot of `/Library/Developer/CommandLineTools/SDKs/MacOSX14.sdk/usr/include`. However, your `18.1.2` version seems to have been configured and built from scratch without any default sysroot, hence why it can't find the right headers.

Long story short, on Apple platforms headers are located in a SDK which is located at `$(xcrun --show-sdk-path)/usr/include`. You can get those headers by using `-isysroot $(xcrun --show-sdk-path)` or by simply running `xcrun clang++ <...>` which will set the `SDKROOT` environment variable before running `clang++`, which lets it set `-isysroot` appropriately.

Closing since I don't think this is truly a regression.
